packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts

```diff
@@ -0,0 +1,133 @@
+ * to make sure all block types are handled and future blocks types will generate compile time errors.
+ */
+type BlockKind = KeysOfUnion<Ogmios.Schema.Block>;
+type OgmiosBlockType =
+  | Ogmios.Schema.BlockAllegra
+  | Ogmios.Schema.BlockAlonzo
+  | Ogmios.Schema.BlockBabbage
+  | Ogmios.Schema.BlockByron
+  | Ogmios.Schema.BlockMary
+  | Ogmios.Schema.BlockShelley;
+
+interface Block<B extends OgmiosBlockType, T extends BlockKind> {
+  block: B;
+  kind: T;
+}
+
+type BlockAndKind =
+  | Block<Ogmios.Schema.BlockAllegra, 'allegra'>
+  | Block<Ogmios.Schema.BlockAlonzo, 'alonzo'>
+  | Block<Ogmios.Schema.BlockBabbage, 'babbage'>
+  | Block<Ogmios.Schema.BlockByron, 'byron'>
+  | Block<Ogmios.Schema.BlockMary, 'mary'>
+  | Block<Ogmios.Schema.BlockShelley, 'shelley'>;
+
+/**
+ * @returns {BlockAndKind} that unlocks type narrowing in switch statements based on `kind`.
+ *          Another advantage of using switch is using exhaustive check of case branches, making sure that future new
+ *          block kinds will cause compilation errors, instead of silently fail or runtime errors.
+ */
+// eslint-disable-next-line complexity
+const getBlockAndKind = (block: Ogmios.Schema.Block): BlockAndKind => {
+  let propName: BlockKind = 'alonzo';
+  if (Ogmios.isAllegraBlock(block)) propName = 'allegra';
+  if (Ogmios.isAlonzoBlock(block)) propName = 'alonzo';
+  if (Ogmios.isBabbageBlock(block)) propName = 'babbage';
+  if (Ogmios.isByronBlock(block)) propName = 'byron';
+  if (Ogmios.isMaryBlock(block)) propName = 'mary';
+  if (Ogmios.isShelleyBlock(block)) propName = 'shelley';
+
+  // If it complains because a branch is not handled, please add logic for the new block type.
+  switch (propName) {
+    case 'allegra':
+      return { block: (block as Ogmios.Schema.Allegra).allegra, kind: 'allegra' };
+    case 'alonzo':
+      return { block: (block as Ogmios.Schema.Alonzo).alonzo, kind: 'alonzo' };
+    case 'babbage':
+      return { block: (block as Ogmios.Schema.Babbage).babbage, kind: 'babbage' };
+    case 'byron':
+      return { block: (block as Ogmios.Schema.Byron).byron, kind: 'byron' };
+    case 'mary':
+      return { block: (block as Ogmios.Schema.Mary).mary, kind: 'mary' };
+    case 'shelley':
+      return { block: (block as Ogmios.Schema.Shelley).shelley, kind: 'shelley' };
+    default: {
+      // will fail at compile time if not all branches are handled
+      // eslint-disable-next-line sonarjs/prefer-immediate-return
+      const _exhaustiveCheck: never = propName;
+      return _exhaustiveCheck;
+    }
+  }
+};
+
+const getBlockBodyHash = (block: Ogmios.Schema.Block): Cardano.BlockId => {
+  const b = getBlockAndKind(block);
+  switch (b.kind) {
+    case 'byron': {
+      return Cardano.BlockId(b.block.hash);
+    }
+    case 'allegra':
+    case 'alonzo':
+    case 'babbage':
+    case 'mary':
+    case 'shelley': {
+      // REVIEW: what to do if blockHash is undefined
```

Comment:
@rhyslbw why could this and [this](https://github.com/input-output-hk/cardano-js-sdk/pull/450/files#diff-89f29146c7c243b9b379062c58515610e8d5decbc534aab0f118cb9cfbefcf64R107) be undefined?

Reply:
We can actually cast `slot` to a non-null property for anything later than Byron. There was a quirk with the original era that meant particular blocks do not have a slot (or block number) :upside_down_face: Since Shelley, all blocks are created within slots and have a number :lotus_position:

Reply:
[1e9a079](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1e9a0799abcaf1cad8e6e8f49bee60ec558edc95)